### PR TITLE
Evolution to GALAHAD 5

### PR DESCRIPTION
**Pull request by amontoison**
Nick: These are all of the preparatory changes for GALAHAD 5. The new edition has new interfaces to julia, and supports the optional multprecision HSL subset, along with other less significant changes.

### CODE REVIEW

**Comment by amontoison** [comment]
@nimgould You forgot to push `hsl_subset.h`.

**Comment by nimgould** [comment]
I added it. Did I need to? I am a bit lost as to where things are, what is mine and what has been created by you

**Comment by amontoison** [comment]
> @nimgould You forgot to push `hsl_subset.h`.

I just created an empty `meson.build` file in `src/metis`.
You can click directly on this commit here to see what I modified but it's almost nothing.

**Comment by nimgould** [comment]
I can no longer commit, I get 

nothing added to commit but untracked files present (use "git add" to track)
To https://github.com/ralna/GALAHAD
 ! [rejected]          galahad5-nick -> galahad5-nick (non-fast-forward)
error: failed to push some refs to 'https://github.com/ralna/GALAHAD'
hint: Updates were rejected because the tip of your current branch is behind
hint: its remote counterpart. Integrate the remote changes (e.g.
hint: 'git pull ...') before pushing again.
hint: See the 'Note about fast-forwards' in 'git push --help' for details.

I tried a git pull, but that didn't help.

Can we get rid of the old galahad5, or are we keeping that for anything?

**Comment by nimgould** [comment]
git status says

On branch galahad5-nick
Your branch and 'origin/galahad5-nick' have diverged,
and have 1 and 1 different commits each, respectively.
  (use "git pull" to merge the remote branch into yours)

Last command done (1 command done):
   pick 448dac29 Evolution to GALAHAD 5
No commands remaining.
You are currently editing a commit while rebasing branch 'galahad5' on '82e7fa91'.

So I seem to be on galahad5-nick but it is complaining about a rebase to galahad5. How do I stop the rebase, and is it this that is coonfusing everything??

**Comment by nimgould** [comment]
% git pull
fatal: It seems that there is already a rebase-merge directory, and
I wonder if you are in the middle of another rebase.  If that is the
case, please try
	git rebase (--continue | --abort | --skip)
If that is not the case, please
	rm -fr ".git/rebase-merge"
and run me again.  I am stopping in case you still have something
valuable there.

**Comment by nimgould** [comment]
(PS I tried to fix the broken galahad_modules.h, but it won't accept the push)

**Comment by amontoison** [comment]
`git rebase --abort`
`git fetch origin`
`git rebase origin/galahad5-nick`

(Fix the conflicts and add the file:
`git add include/galahad_modules.h`

git rebase --continue
git push

**Comment by nimgould** [comment]
Git is really broken. I follow your instructions, update all of the compromised files, git rebase --continue, but then at the next commit, all the files I updated have conflicts reinserted. How do I get out of this endless loop?

**Comment by nimgould** [comment]
Despite using the git rebase origin/galahad5-nick command, it  keeps doing things to galahad5 (no nick), and then complains that all of the changes I made are now conflicted again. I've been through the same cycle of fixing conflicts  four times now, and I cannot see how this will ever converge

**Comment by nimgould** [comment]
This was my mistake, I missed the git checkout galahad5-nick, and all is now fine. How do I get it to run the actions?

**Comment by amontoison** [comment]
> This was my mistake, I missed the git checkout galahad5-nick, and all is now fine. How do I get it to run the actions?

`git push`

**Comment by nimgould** [comment]
Thank you. I thought I'd done that, but obviously not.

For some reason the build gives

./src/external/hsl/hsl_ma48/hsl_ma48r.f90:139:11:
737

738
  139 |       USE GALAHAD_SYMBOLS
739
      |           1
740
Fatal Error: Cannot open module file ‘galahad_symbols.mod’ for reading at (1): No such file or directory
 
But the earlier 

[656/2783] Compiling Fortran object libgalahad_single.so.p/meson-generated_single_symbols.f90.o

should provide this module file. Is something missing in meson?

**Comment by nimgould** [comment]
That was my blunder, a programming bug that wasn't picked up locally.

**Comment by nimgould** [comment]
Another one, this looks like a meson issue, This is supposed to be a 32bit integer single compile, but the -DINTEGER_64 flag is being passed to the compiler

gfortran -Ilibgalahad_single.so.p -I. -I.. -Iinclude -I../include -I../src/dum/include -I../src/metis/include -Isrc/ampl -I../src/ampl -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -O0 -g -fopenmp -fPIC -DSINGLE -DGALAHAD_BLAS -DGALAHAD_LAPACK -DDUMMY_SMUMPS -DDUMMY_MKL_PARDISO -DDUMMY_PARDISO -DDUMMY_PASTIXF -DDUMMY_SPMF -DDUMMY_WSMP -DDUMMY_MPI -DDUMMY_HSL -DINTEGER_64 -Jlibgalahad_single.so.p -o libgalahad_single.so.p/meson-generated_single_uls.f90.o -c libgalahad_single.so.p/single_uls.f90
817
../src/uls/uls.F90:44:10:
818
   44 |      USE hsl_ma48_precision
819
      |          1
820
Fatal Error: Cannot open module file ‘gal_hsl_ma48_single_64.mod’ for reading at (1): No such file or directory

**Comment by amontoison** [comment]
Nick, you looked at a 64-bit integer build.

But we have a similar issue with a 32-bit integer build, we have something wrong the preprocessed macros for HSL:
```
https://github.com/ralna/GALAHAD/actions/runs/8633938903/job/23668183354?pr=272#step:12:825
```

**Comment by nimgould** [comment]
How do we find what the compiler did when compiling hsl_ma57r.f90? what name was the resulting .mod file?

**Comment by amontoison** [comment]
@nimgould I found the culprit. It's because with `galahad_modules.h`, I add the prefix `gal` with the dummy modules.
I added a bunch of macros in `hsl_subset.h` to that.
I think that we should merge some header files together.
It starts to get hard to maintain with all these macros in different files.

**Comment by amontoison** [comment]
New error:
```shell
 gfortran -Ilibgalahad_double.so.p -I. -I.. -Iinclude -I../include -I../src/dum/include -I../src/metis/include -Isrc/ampl -I../src/ampl -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -O0 -g -fopenmp -fPIC -DDOUBLE -DGALAHAD_BLAS -DGALAHAD_LAPACK -DDUMMY_DMUMPS -DDUMMY_MKL_PARDISO -DDUMMY_PARDISO -DDUMMY_PASTIXF -DDUMMY_SPMF -DDUMMY_WSMP -DDUMMY_MPI -DDUMMY_HSL -Jlibgalahad_double.so.p -o libgalahad_double.so.p/meson-generated_double_blas_interface.f90.o -c libgalahad_double.so.p/double_blas_interface.f90
../src/lapack/blas_interface.F90:40:10:
   40 |        DOUBLE PRECISION :: DNRM2
      |          1
Error: Unclassifiable statement at (1)
../src/lapack/blas_interface.F90:51:31:
   51 |        SUBROUTINE SROTG( a, b, c, s )
      |                               1
......
   57 |        SUBROUTINE DROTG( a, b, c, s )
      |                               2
Error: Ambiguous interfaces in generic interface 'rotg' for ‘galahad_srotg’ at (1) and ‘galahad_drotg’ at (2)
```

**Comment by nimgould** [comment]
Thanks for finding the include issue. I realised that it must be something like this in the middle of the night!

Yes, we should simplify the hsl header files. The main complication is, of course, the dummy names you need for your build.

I'll see what the blas_interface problem is and fix it.

And, yes, it should be trivial to remove the explicit zd11. I will look

Thanks for your help on this

**Comment by nimgould** [comment]
Unfortunately your fix to hsl_subset.h broke things here! I will recover and do it again.

File: include/hsl_subset.h [hunk 1]
```diff
@@ -429,8 +642,21 @@
...
+#define KB21AI GALAHAD_KB21AI_64
+#define KB21BI GALAHAD_KB21BI_64
+#define KB21CI GALAHAD_KB21CI_64
+#define KB21DI GALAHAD_KB21DI_64
+#define KB21EI GALAHAD_KB21EI_64
+#define KB21FI GALAHAD_KB21FI_64
+#define KB21GI GALAHAD_KB21GI_64
+#define KB21HI GALAHAD_KB21HI_64
+#define kb07ai galahad_kb07ai_64
+#ifdef DUMMY_HSL
```
Comment (amontoison) [inline include/hsl_subset.h:658]: @nimgould 
I did an error here. This `#ifdef` should be at line 648.

**Comment by nimgould** [comment]
I am working on this at the moment, it is a bit of a mess!

**Comment by nimgould** [comment]
I believe that the issue you reported for blas_interface.F90 is that you are using the -DDOUBLE flag; unfortunately DOUBLE is a reserved keyword in C++; we are careful elsewhere not to use DOUBLE in any of the header files. See what happens if you remove this flag

File: meson.build [hunk 2]
```diff
@@ -253,7 +253,8 @@ if host_machine.system() != 'linux'
 endif
 
 extra_args_single = ['-DSINGLE']
-extra_args_double = ['-DDOUBLE']
+#extra_args_double = ['-DDOUBLE']
+extra_args_double = ['']
```
Comment (amontoison) [inline meson.build:257]:
```suggestion
extra_args_double = []
```

**Comment by amontoison** [comment]
> I believe that the issue you reported for blas_interface.F90 is that you are using the -DDOUBLE flag; unfortunately DOUBLE is a reserved keyword in C++; we are careful elsewhere not to use DOUBLE in any of the header files. See what happens if you remove this flag

It fixed the issue with the.BLAS interface.
I checked CI builds and we can almost compile.
We have an error during linking but it's because of the preprocessed symbols (HSL / METIS).
I can fix it later today.

**Comment by nimgould** [comment]
I forgot to add a couple of local metis files, I have them at home and will commit from there. The last actions complained about not having mc23b, I need to check, again later

**Comment by amontoison** [comment]
We have the following error with the dummy `hsl_ma57r` (64-bit integer):
```
FAILED: libgalahad_single.so.p/meson-generated_single_ma27r.f.o 
gfortran -Ilibgalahad_single.so.p -I. -I.. -Iinclude -I../include -I../src/dum/include -I../src/metis/include -Isrc/ampl -I../src/ampl -fdiagnostics-color=always -D_FILE_OFFSET_BITS=64 -O0 -g -fopenmp -fPIC -DSINGLE -DGALAHAD_BLAS -DGALAHAD_LAPACK -DDUMMY_SMUMPS -DDUMMY_MKL_PARDISO -DDUMMY_PARDISO -DDUMMY_PASTIXF -DDUMMY_SPMF -DDUMMY_WSMP -DDUMMY_MPI -DDUMMY_HSL -DINTEGER_64 -Jlibgalahad_single.so.p -o libgalahad_single.so.p/meson-generated_single_ma27r.f.o -c libgalahad_single.so.p/single_ma27r.f
../src/external/hsl/ma27/ma27r.f:69:70:

   69 |       INTEGER ( KIND = ip_ ), INTENT( IN ), DIMENSION( n, 3 ) :: IKEEP
      |                                                                      1
Error: Symbol at (1) is not a DUMMY variable
../src/external/hsl/ma27/ma27r.f:65:68:

   65 |       INTEGER ( KIND = ip_ ), INTENT( IN ) :: n, nz, la, liw, nsteps
      |                                                                    1
Error: Symbol at (1) is not a DUMMY variable
../src/external/hsl/ma27/ma27r.f:111:71:

  111 |       INTEGER ( KIND = ip_ ), INTENT( OUT ), DIMENSION( nsteps ) :: IW1
      |                                                                       1
Error: Symbol at (1) is not a DUMMY variable
../src/external/hsl/ma27/ma27r.f:109:72:

  109 |       INTEGER ( KIND = ip_ ), INTENT( IN ) :: n, la, liw, maxfrt, nsteps
      |                                                                        1
Error: Symbol at (1) is not a DUMMY variable
../src/external/hsl/ma27/ma27r.f:188:68:

  188 |       INTEGER ( KIND = ip_ ), INTENT( IN ), DIMENSION( nblk ) :: IW2
      |                                                                    1
Error: Symbol at (1) is not a DUMMY variable
../src/external/hsl/ma27/ma27r.f:186:70:

  186 |       INTEGER ( KIND = ip_ ), INTENT( IN ) :: n, la, liw, maxfnt, nblk
      |                                                                      1
Error: Symbol at (1) is not a DUMMY variable
```

**Comment by nimgould** [comment]
I suspect that this is because a previous line overlowed the 72 character limit after one of your substitutions. For instance
       SUBROUTINE GALAHAD_MA27BRD_64 n, nz, IRN, ICN, A, la, IW, liw, IKEEP,
is 75 chars long. I have substituted GALAHAD_ to GAL_ in hsl_subset.h. Let's see ...

**Comment by amontoison** [comment]
@nimgould You just forgot some dummy HSL routines:
https://github.com/ralna/GALAHAD/pull/272/commits/c1d689b5c6b25a1dbbace07e973c1f7becd51809

**Comment by nimgould** [comment]
/home/runner/work/GALAHAD/GALAHAD/builddir_int32/../src/gls/gls.F90:1295: undefined reference to `gal_mc21b_'
(etc). meson doesn't seem to have compiled mc21r.f90, although it is in the meson.build

**Comment by amontoison** [comment]
> /home/runner/work/GALAHAD/GALAHAD/builddir_int32/../src/gls/gls.F90:1295: undefined reference to `gal_mc21b_' (etc). meson doesn't seem to have compiled mc21r.f90, although it is in the meson.build

Nick, you also need to update macros in `hsl_subset.h`.
You still use the old prefix `GALAHAD_` in this file.

**Comment by nimgould** [comment]
A number of the issues seem to relate to the dummy metis. GALAHAD no longer uses a metis library, it compiles its own version so that it knows that the arguments are right. We have deliberatly changed the names of all of our metis components (and for the make build, the library itself so not to clash).  Thus we need meson to compile everything in the ./src/metis library not dummy_metis. In addition the -DMETIS4 flag should not be set (I really should remove this possible preprocess). If any external package wants to use another metis, that is fine as the symbols will differ. As it stands, MUMPS has issues with metis when it links the wrong version, as you mentioned.

**Comment by amontoison** [comment]
> A number of the issues seem to relate to the dummy metis. GALAHAD no longer uses a metis library, it compiles its own version so that it knows that the arguments are right. We have deliberatly changed the names of all of our metis components (and for the make build, the library itself so not to clash). Thus we need meson to compile everything in the ./src/metis library not dummy_metis. In addition the -DMETIS4 flag should not be set (I really should remove this possible preprocess). If any external package wants to use another metis, that is fine as the symbols will differ. As it stands, MUMPS has issues with metis when it links the wrong version, as you mentioned.

I noticed that you added the source code of METIS.

The issue is that you removed all the previous files in src/metis (like meson.build and the dummy file), but you didn't add a fix for the Meson build system. As a result, it's broken because we can't compile it. I temporarily reintroduced the dummy file, but it won't work anymore since we need additional routines, correct?

Is METIS4 still supported? I noticed the macro -DMETIS4, so I assumed it was still supported.

Nick, you can't modify 781 files in one commit without explaining the main modifications :smiling_face_with_tear:.

**Comment by amontoison** [comment]
> Yup, and I just updated those too

Sorry wrong commit, you forgot to implement the dummy routines: https://github.com/ralna/GALAHAD/pull/272/commits/23a92b732248509a6224b9f2a5b79ef0e5a611f4

@nimgould, what are the difference between `hsl_metis` and `galahad_metis` should we only use one?

**Comment by amontoison** [comment]
> > > A number of the issues seem to relate to the dummy metis. GALAHAD no longer uses a metis library, it compiles its own version so that it knows that the arguments are right. We have deliberatly changed the names of all of our metis components (and for the make build, the library itself so not to clash). Thus we need meson to compile everything in the ./src/metis library not dummy_metis. In addition the -DMETIS4 flag should not be set (I really should remove this possible preprocess). If any external package wants to use another metis, that is fine as the symbols will differ. As it stands, MUMPS has issues with metis when it links the wrong version, as you mentioned.
> > 
> > 
> > I noticed that you added the source code of METIS.
> > The issue is that you removed all the previous files in src/metis (like meson.build and the dummy file), but you didn't add a fix for the Meson build system. As a result, it's broken because we can't compile it. I temporarily reintroduced the dummy file, but it won't work anymore since we need additional routines, correct?
> 
> Yes, I did write a meson.build for the metis directory, but I presume I forgot to add it, and it seems to be blank now (I could have dreamed this!).
> 
> Effectively one has to compile everything mentioned in the SRC variable in the makemaster file plus adapter.c and metis.f90. Any static linking is against -lmetis_nodend or -lmetis64_nodend
> 
> > Is METIS4 still supported? I noticed the macro -DMETIS4, so I assumed it was still supported.
> 
> No, I only left this for the time being, but it should be removed.
> 
> > Nick, you can't modify 781 files in one commit without explaining the main modifications 🥲.
> 
> Clearly I can as I did :) I am completely exhausted by this all, and haven't any energy left to write another encyclopedia!! The entire rewrite will be the focus of a report at some stage.

Thanks Nick!

Okay, take some rest.
I will debug the latest errors.
We're getting there.

**Comment by nimgould** [comment]
Ooops, yes, I forgot to add the mc13br and mc23er, Strictly, these need to have the full arguments exposed as they are called by sls, I'll do that. 

Ignore the hsl_metis directory, al that is required is the symlink to metis. So I suppose it is just galahad_metis that we need, and we need to map hsl_metis to galahad_metis in hsl_subset.h

Sorry, got to go on vacation!! over and out for a couple of weeks

**Comment by amontoison** [comment]
> Ooops, yes, I forgot to add the mc13br and mc23er, Strictly, these need to have the full arguments exposed as they are called by sls, I'll do that.

:+1: 

> Ignore the hsl_metis directory, al that is required is the symlink to metis. So I suppose it is just galahad_metis that we need, and we need to map hsl_metis to galahad_metis in hsl_subset.h

I removed the folder `src/external/hsl/hsl_metis` and only use the routine `galahad_metis` now.
I also updated `hsl_subset.h` to keep these macros:
```c
#ifdef INTEGER_64
#define hsl_metis galahad_metis_64
#else
#define hsl_metis galahad_metis
#endif
```

I can compile again GALAHAD on my laptop :muscle: 
 
> Sorry, got to go on vacation!! over and out for a couple of weeks

Enjoy your holidays! :beach_umbrella:

File: src/metis/error.c [hunk 3]
```diff
@@ -125,69 +125,71 @@ int gk_siguntrap()
...
-  
+
 
 /*************************************************************************/
 /*! \brief Thread-safe implementation of strerror() */
 /**************************************************************************/
 char *gk_strerror(int errnum)
+/* comment this out as it is more trouble than it is worth! */
+/* { */
+/* #if defined(WIN32) || defined(__MINGW32__) */
```
Comment (amontoison) [inline src/metis/error.c:178]: @nimgould 
The macro that is compiler independent on Windows is `_WIN32`.

Comment (nimgould) [reply src/metis/error.c:178]: Understood. This part came from the metis people, and I suspect that they defined _WIN32 somewhere else. Unfortunately, the whole thing was out of date, and wouldn't work on macs, so I decided simply to go for the thread-unsafe version.

**Comment by amontoison** [comment]
@nimgould Do we need to install an external package for `-lregex` on Windows?
I think that we should provide METIS 5.1.0 and not METIS 5.2.0.
Nothing was tested with this recent release...

**Review by nimgould — COMMENTED** [review]
Excellent. Might I humbly suggest that we also mention the integrated make build system for the whole of the optrove eco-system, since this provides scripts to run cutest examples directly from galahad (and other packages)

**Comment by amontoison** [comment]
> Excellent. Might I humbly suggest that we also mention the integrated make build system for the whole of the optrove eco-system, since this provides scripts to run cutest examples directly from galahad (and other packages)

Yes, It's what we need to test CUTEst examples and the MATLAB interface.
I just prepared a skeleton of `makemaster.yml` for you.
I'm writing an email to you and Jari

**Comment by nimgould** [comment]
OK, I am just pushing a slight addition to Jari's readme to make this point (and to tell anyone who is interested what to do).

**Comment by nimgould** [comment]
Hurrah, all the tests now pass. Are we ready to go ... it's a public holiday here on Monday, nut perhaps Tuesday is the day to merge galahad5-nick to master, and publicize the launch

**Comment by nimgould** [comment]
I spoke too soon ,,, it was the activities on the main branch that succeeded, the same Windows ones on 5-nick still fail. My mistake :(

**Comment by amontoison** [comment]
Once I fix the Windows builds, we can merge the branch `galahad5-nick`.

**Comment by amontoison** [comment]
@nimgould @jfowkes 

I fixed the compilation on Windows. :muscle:  
I spent my afternoon discovering that `GKlib` provides its own `gkregex.c`, and we need to compile it with `-DUSE_GKREGEX`.
Nick didn't include all the files from GKlib in GALAHAD, and this one was needed for Windows.

The Windows builds are still failing, but it's just a few errors in the tests that should be easy to fix.
For 64-bit integers, the errors are just related to `dgo`.
For 32-bit integers, we have 9 tests that are failing, so we should fix them before the release of 5.0.0.
I propose to merge this PR and continue fixing the last details on `master`.

```shell
----------------------------------- stderr -----------------------------------
At line 2232 of file ../src/nls/nls.F90 (unit = 538976288)
Fortran runtime error: Cannot open file '': No such file or directory
```
```shell
----------------------------------- stderr -----------------------------------
At line 3391 of file ../src/llsr/llsr.F90
Fortran runtime error: Unit number is negative and unit was not already opened with OPEN(NEWUNIT=...)
```

**Comment by jfowkes** [comment]
> @nimgould @jfowkes
> 
> I fixed the compilation on Windows. 💪 I spent my afternoon discovering that `GKlib` provides its own `gkregex.c`, and we need to compile it with `-DUSE_GKREGEX`. Nick didn't include all the files from GKlib in GALAHAD, and this one was needed for Windows.
> 
> The Windows builds are still failing, but it's just a few errors in the tests that should be easy to fix. For 64-bit integers, the errors are just related to `dgo` tests (even if the failure is in hash.F90).
> 
> ```shell
> At line 416 of file ../src/hash/hash.F90 (unit = 24576)
> Fortran runtime error: Unit number in I/O statement too large
> ```
> 
> For 32-bit integers, we have 9 tests that are failing, so we should fix them before the release of 5.0.0. I propose to merge this PR and continue fixing the last details on `master`.
> 
> ```shell
> ----------------------------------- stderr -----------------------------------
> At line 2232 of file ../src/nls/nls.F90 (unit = 538976288)
> Fortran runtime error: Cannot open file '': No such file or directory
> ```
> 
> ```shell
> ----------------------------------- stderr -----------------------------------
> At line 3391 of file ../src/llsr/llsr.F90
> Fortran runtime error: Unit number is negative and unit was not already opened with OPEN(NEWUNIT=...)
> ```

@amontoison agreed this PR is now a real monster! I also suggest we merge the galahad5-nick branch to master and fix the last few test errors there, how does that sound @nimgould ?

**Comment by nimgould** [comment]
Sorry for the delay ...

I think I have fixed the issues you mentioned for nls and llsr. However, the dgo one has stumped me, as the unit number is initialized to 6, and not changed anywhere. Thus it must be an overwrite issue. Trouble is, no compiler I have says that there is an issue, and I have no access to a Windows machine. Help!

Yes, @amontoison it is time for the mega-merge. I would be grateful if you could do it, I won't make any more changes to  galahad5-nick, and I then switch branches once you are done. Thank you

**Comment by amontoison** [comment]
@jfowkes You just force-pushed and removed the last two commits of @nimgould.

**Comment by jfowkes** [comment]
@amontoison so sorry! I've asked @nimgould to force push his branch and fix it...

**Comment by jfowkes** [comment]
> What does that mean I have to do? Instructions, please

Please do `git push -f` from within the GALAHAD folder.

**Comment by nimgould** [comment]
I got an Everything up-to-date
Have I lost stuff?

**Comment by jfowkes** [comment]
No it's still here:
https://github.com/ralna/GALAHAD/commit/be3c7c8de3f700dab405ae07e2da0cea94f33df7
https://github.com/ralna/GALAHAD/commit/64e0850ea5ef8af8b6860f11bb033df3156612da
Just not on this branch, @amontoison how can we best fix this?

**Comment by amontoison** [comment]
> I got an Everything up-to-date Have I lost stuff?

Please don't do a `git pull`.

**Comment by nimgould** [comment]
I probably already did!. 
Git told me to!

**Comment by amontoison** [comment]
It's not your fault Nick. Jari is the culprit!

**Comment by nimgould** [comment]
Yup, I've lost all of the changes I made .... of course those https://github.com/ralna/GALAHAD/commit/be3c7c8de3f700dab405ae07e2da0cea94f33df7
https://github.com/ralna/GALAHAD/commit/64e0850ea5ef8af8b6860f11bb033df3156612da
messages say what the changes were, so in the worst case I could try to redo them

**Comment by nimgould** [comment]
Is it not possible to tell giit to reapply those ones above?

**Comment by amontoison** [comment]
Can you try this command Nick?
```shell
git checkout 64e0850ea5ef8af8b6860f11bb033df3156612da
```
If it works then:
```shell
git checkout -b nick-backup
git push
```

**Comment by nimgould** [comment]
The checkout seems to have all the changes I made. Right, I'll try your two commands. Wish me luck ...

**Comment by amontoison** [comment]
If it's not working, we can still go to the page
https://github.com/ralna/GALAHAD/tree/64e0850ea5ef8af8b6860f11bb033df3156612da
and download the files modified by your two commits one by one, Nick.

Fortunately, GitHub keeps a copy of these commits somewhere.

**Comment by nimgould** [comment]
% git checkout -b nick-backup
Switched to a new branch 'nick-backup'
% git push
fatal: The current branch nick-backup has no upstream branch.
To push the current branch and set the remote as upstream, use

    git push --set-upstream origin nick-backup

Should I do that?

**Comment by amontoison** [comment]
> % git checkout -b nick-backup Switched to a new branch 'nick-backup' % git push fatal: The current branch nick-backup has no upstream branch. To push the current branch and set the remote as upstream, use
> 
> ```
> git push --set-upstream origin nick-backup
> ```
> 
> Should I do that?

Yes!!!

**Comment by nimgould** [comment]
Done. It looks ok 

% git push --set-upstream origin nick-backup
Enumerating objects: 64, done.
Counting objects: 100% (64/64), done.
Delta compression using up to 16 threads
Compressing objects: 100% (38/38), done.
Writing objects: 100% (38/38), 8.39 KiB | 8.39 MiB/s, done.
Total 38 (delta 30), reused 2 (delta 0), pack-reused 0
remote: Resolving deltas: 100% (30/30), completed with 25 local objects.
remote: 
remote: Create a pull request for 'nick-backup' on GitHub by visiting:
remote:      https://github.com/ralna/GALAHAD/pull/new/nick-backup
remote: 
To https://github.com/ralna/GALAHAD
 * [new branch]        nick-backup -> nick-backup
Branch 'nick-backup' set up to track remote branch 'nick-backup' from 'origin'

**Comment by amontoison** [comment]
Yep yep

**Comment by nimgould** [comment]
Thank you so much, you are my hero of the hour, perhaps day

**Comment by amontoison** [comment]
You're welcome Nick!
I have earned the right to correct my copies now...

**Comment by nimgould** [comment]
Of course, of course, you can correct as much as you like. Now I've lost that beer in the excitement

**Comment by jfowkes** [comment]
Thank you both and please accept my sincere apologies. I have now fixed the README on the master branch.

**Comment by jfowkes** [comment]
@amontoison can we kill the GitHub Pages build? No idea why that's running on master, docs are hosted from https://github.com/ralna/galahad_docs

**Comment by amontoison** [comment]
> Thank you both and please accept my sincere apologies. I have now fixed the README on the master branch.

If you pay me a beer at Montréal during your visit for ISMP, I accept your apologies :stuck_out_tongue:

**Comment by jfowkes** [comment]
> > Thank you both and please accept my sincere apologies. I have now fixed the README on the master branch.
> 
> If you pay me a beer at Montréal during your visit for ISMP, I accept your apologies 😛

You're on, one of those famous American Pale Ales of yours 😜

**Comment by amontoison** [comment]
> @amontoison can we kill the GitHub Pages build? No idea why that's running on master, docs are hosted from https://github.com/ralna/galahad_docs

I think that I killed the build. Let's see when a new commit will be added on `master`.

**Comment by jfowkes** [comment]
> > @amontoison can we kill the GitHub Pages build? No idea why that's running on master, docs are hosted from https://github.com/ralna/galahad_docs
> 
> I think that I killed the build. Let's see when a new commit will be added on `master`.

Thanks @amontoison. @nimgould we seem to have an old stray `index.html`in the root GALAHAD folder, I assume this was from the time when the GALAHAD website was planned to be hosted from this repository but is no longer in use?

**Comment by nimgould** [comment]
Yes, indeed. Infact it is only a week old, it was from my failed attempt to host the fortran pdf docs from the galahad site. Fortunately, following Jari's suggestion, these are now on the galahad_docs site

**Comment by nimgould** [comment]
It will be gone at my next commit

**Pull request merged**